### DISCUSSION TITLE
Against structures:metadata

### DISCUSSION THREAD

**Discussion by iamdrscott** [niemopen/ntac-admin#66] [General]
The [Attribute augmentation and metadata replacement](https://github.com/niemopen/ntac-admin/blob/main/examples/src/Augmentation/README.md) paper proposes *reference attributes* as the way to augment simple content with an element (06-AugSCwithE).  For example:

```
  <nc:PersonEducation>
    <nc:EducationDescriptionText my:privacyAssertionRef="p01">PhD</nc:EducationDescriptionText>
    <nc:EducationInProgressIndicator>false</nc:EducationInProgressIndicator>
  </nc:PersonEducation>
```

Instead of introducing reference attributes as a new NIEM thing, we could continue to use `structures:metadata` for this.

```
  <nc:PersonEducation>
    <nc:EducationDescriptionText structures:metadata="p01">PhD</nc:EducationDescriptionText>
    <nc:EducationInProgressIndicator>false</nc:EducationInProgressIndicator>
  </nc:PersonEducation>
```

Among other things, this would not be a breaking change.  But I am going to argue that we *want* a breaking change.  We should take this opportunity to rid ourselves of the metadata mechanism.  Why?  Because the metadata *mechanism* is not really about representing "data about data".  The *mechanism* is actually just another form of augmentation.  Consider:

* *I have some data about my data.  Do I have to use `structures:metadata` for this?*  Well, no, you aren't required to use `structures:metadata` for data about your data, and everyone who can possibly avoid it does so. In NIEM, data about your data is just more data, and In your namespaces you can model this however you want.
* *I need to add an element to this simple content in someone else's namespace, but this added element really isn't "data about data".  Can I still use `structures:metadata` for this?*  Well, sure, you can do that, and it will work just fine.

You don't have to use `structures:metadata` for "data about data", and you can use `structures:metadata` for stuff that is not "data about data".  That means `structures:metadata` isn't really *about* metadata at all.  It's really just another form of augmentation.  But people naturally *think it's about* metadata.  We waste cycles explaining the mechanism.  They waste cycles coming to the correct understanding.  So let's get rid of the metadata mechanism, just have augmentation, and save those cycles.

We should keep `nc:Metadata` in the model.  Message designers can still use it to augment components in someone elses' namespace.  But now they can also use it as an ordinary child in complex content in their own namespace, which is what almost everyone will want to do.

If we want to avoid two ways of augmenting complex content with an element, we can write a rule to say you can't have a reference attribute on a type with child elements.

**Comment by cdmgtri** [comment]
Documenting some of the discussion from last week...

**Alternative to simple content augmentation**

- Continue using `structures:metadata` (or `structures:relationshipMetadata`) for simple content, but relax the requirements
- Do not require refs to ids of official NIEM metadata properties
  - Can ref something like `nc:Metadata`
  - Could also ref individual properties like `nc:ExpirationDate` and `nc:ReporterPerson`
- Optional or required to add appinfo to a property or type declaration describing what kind of refs the metadata attribute should contain
- Remove everything else about metadata from the NDR

**Comparison**

Simple content | Augmentation | `structures:metadata` / `structures:relationshipMetadata`
-------------- | ------------ | ---------------------
Impact vs 5.0 | Breaking | Non-breaking
Difficulty for new users | Harder: User must build custom attribute ref alternates for existing properties | Easier: Bucket for refs already exists
NIEM-ness | NDR doesn't allow xs:IDREF or xs:IDREFs outside of structures | NIEM doesn't support generic buckets of random stuff in other places in the model
Readability | Better: One custom IDREF attribute per requirement plus more semantics in name |
Machine validation | No automatic validation that refs point to correct elements.  Ref attribute namespace doesn't necessarily match the original element namespace.  Additional appinfo may be required to build Schematron rules. |  No automatic validation that ref points to correct elements.  Appinfo could be used to build Schematron rules.
Granularity (new) | For REF or EXT schemas, still won't be very granular.  If you want to augment one text property, the new attribute will be available for all of them.  Same issue as different person profiles in the same message. | Already available on all elements.  All or nothing (either keep it in structures or remove it).  Could subset it for advanced users like the augmentation approach - remove it from structures and add it to specific types via the attribute wildcards.
Subsets (new) | Augmenting simple content from NIEM types requires modifying subsets.  If done outside of CMF / CMF Tool, changes will be lost and need to be redone each time a subset is regenerated. <br /> <br /> In some cases, we could go ahead and add these directly in the model (e.g., add `cui:portionMarkingMetataRef` and `nc:metadataRef` to `nc:TextType`) and this could be wired in for the users automatically via the SSGT. | Does not require modifications to subsets.
